sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -17,161 +17,565 @@
+      responseStream = client.readRows(request);
+      responseIterator = responseStream.iterator();
+      rowIterator = null;
+      return readNextRow();
     }
 
     @Override
-    public boolean advance() {
-      if (rowIterator == null || !rowIterator.hasNext()) {
+    public boolean advance() throws IOException {
+
+      currentPosition =
+          currentPosition.toBuilder().setOffset(currentPosition.getOffset() + 1).build();
+
+      while (true) {
+        if (!readNextRow()) {
+          return rangeTracker.markDone();
+        }
+
+        if (rangeTracker.tryReturnRecordAt(true, currentPosition)) {
+          return true;
+        }
+
+        if (rangeTracker.isDone()) {
+          return false;
+        }
+
+        if (responseStream != null) {
+          responseStream.cancel();
+          responseStream = null;
+        }
+
+        StreamPosition splitPosition =
+            StreamPosition.newBuilder()
+                .setStream(currentPosition.getStream())
+                .setOffset(currentPosition.getOffset() - 1)
+                .build();
+
+        currentPosition = processSplit(splitPosition, SplitDisposition.PRIMARY);
+
+        ReadRowsRequest request =
+            ReadRowsRequest.newBuilder().setReadPosition(currentPosition).build();
+        responseStream = client.readRows(request);
+        responseIterator = responseStream.iterator();
+        rowIterator = null;
+        estimatedStreamLength = StreamOffsetRangeTracker.OFFSET_INFINITY;
+      }
+    }
+
+    private StreamPosition processSplit(
+        StreamPosition splitPosition, SplitDisposition disposition) {
+      checkArgument(disposition != SplitDisposition.SELF);
+
+      SplitReadStreamRequest request =
+          SplitReadStreamRequest.newBuilder().setOriginalStream(splitPosition.getStream()).build();
+      SplitReadStreamResponse response = client.splitReadStream(request);
+
+      checkState(
+          response.hasPrimaryStream(),
+          "Invalid split response [%s] missing primary stream",
+          response);
+      checkState(
+          response.hasRemainderStream(),
+          "Invalid split response [%s] missing remainder stream",
+          response);
+
+      if (response.getPrimaryStream().getRowCount() > splitPosition.getOffset()) {
+
+        //
+        // The split position is contained within the primary stream, so the primary source is the
+        // primary stream ID with the same offset range as the parent source and the residual source
+        // is the entirety of the remainder stream.
+        //
+        // We're in a "split pending" state, so any attempt to return a record, split the range
+        // tracker at an offset, or mark a server-side split as pending has been rejected. This
+        // also means that the current source has not been modified, and can be updated with proper
+        // synchronization.
+        //
+
+        if (disposition == SplitDisposition.PRIMARY) {
```

Comment:
I would like to understand what's unclear from your perspective so that I can add the appropriate context to the comment here. It is true that "data already read by the current reader" + "data read by the current reader after the current